README.md
# drop_down_menu
一个简单的drop_down_menu，支持任意widget。

将JDDropdownMenuWidget放到任意位置即可。



```dart
JDDropdownMenuWidget(
              controller: controller,
              itemStyle: JDDropdownMenuWidgetStyle.style2,
              click: (int index) {
                print('$index');
              },
              items: [
                JDDropdownMenuItem(
                  maxHeight: 300,
                  title: Container(
                    alignment: Alignment.topCenter,
                    margin: EdgeInsets.only(
                      left: 10,
                      right: 10,
                    ),
                    child: Text(title1),
                  ),
                  menu: CityWidget(
                    click: (String selectedValue) {
                      setState(() {
                        title1 = selectedValue;
                        controller.hide();
                      });
                      print('选择的城市为:$selectedValue');
                    },
                  ),
                ),
                JDDropdownMenuItem(
                  maxHeight: 300,
                  title: Container(
                    alignment: Alignment.topCenter,
                    margin: EdgeInsets.only(
                      left: 10,
                      right: 10,
                    ),
                    child: Text(title2),
                  ),
                  menu: MenuListWidget(
                    selectedValue: title2,
                    click: (String value) {
                      setState(() {
                        title2 = value;
                        controller.hide();
                      });
                    },
                  ),
                ),
                JDDropdownMenuItem(
                  title: Container(
                    alignment: Alignment.topCenter,
                    margin: EdgeInsets.only(
                      left: 10,
                      right: 10,
                    ),
                    child: Text('菜单3'),
                  ),
                  menu: Container(
                    color: Colors.blue,
                  ),
                ),
              ],
   )
```



demo地址：https://github.com/JDongKhan/jd_dropdown_menu_demo.git





![demo](./demo.gif)
![demo](https://github.com/JDongKhan/drop_down_menu/blob/main/demo.gif)

![demo](https://github.com/JDongKhan/drop_down_menu/blob/3ad170afc1ab4b82341a51e4fa02fdd8492c627a/demo.gif)







CHANGELOG.md
## [0.0.5] - fix bugs

## [0.0.3] - support null-safety

## [0.0.2] - add README


## [0.0.1] - init project.

* TODO: Describe initial release.


lib/jd_pop_widget.dart
import 'package:flutter/material.dart';

/// @author jd
///

typedef ClickCallBack = void Function(int selectIndex, String selectText);

//支持自定义 pop 菜单

void showPop({
  required BuildContext context,
  required List<String> items,
  double? top,
  double? left,
  double? right,
  double width = 100,
  double cellHeight = 40,
  Color backgroundColor = Colors.white,
  Color textColor = Colors.black,
  Color dividerColor = const Color(0xFFE6E6E6),
  required Color barrierColor,
  bool hiddenArrow = false,
  bool barrierDismissible = true,
  ClickCallBack? clickCallback,
}) {
  Widget _buildMenuLineCell(dataArr) {
    return MediaQuery.removePadding(
      context: context,
      removeTop: true,
      child: ListView.separated(
        itemCount: dataArr.length,
        physics: const NeverScrollableScrollPhysics(),
        itemBuilder: (BuildContext context, int index) {
          return GestureDetector(
              onTap: () {
                Navigator.pop(context);
                if (clickCallback != null) {
                  clickCallback(index, items[index]);
                }
              },
              child: Container(
                height: cellHeight,
                margin: EdgeInsets.only(
                  left: 10,
                ),
                child: Align(
                  alignment: Alignment.centerLeft,
                  child: Text(dataArr[index],
                      style: TextStyle(fontSize: 16, color: textColor)),
                ),
              ));
        },
        separatorBuilder: (context, index) {
          return Divider(
            height: 0.1,
            color: dividerColor,
          );
        },
      ),
    );
  }

  showPopView(
    context: context,
    widget: _buildMenuLineCell(items),
    top: top,
    left: left,
    right: right,
    width: width,
    height: cellHeight * items.length,
    backgroundColor: backgroundColor,
    barrierColor: barrierColor,
    hiddenArrow: hiddenArrow,
    barrierDismissible: barrierDismissible,
  );
}

//
// static void showPopView({
//   @required BuildContext context,
//   Widget widget,
//   double top,
//   double left,
//   double right,
//   double width = 120,
//   double height,
//   Color backgroundColor = Colors.white,
//   bool hiddenArrow = false,
// }) {
//   _buildMenusView() {
//     return Positioned(
//       right: right,
//       left: left,
//       top: top,
//       child: Column(
//         crossAxisAlignment: CrossAxisAlignment.end,
//         children: <Widget>[
//           if (!hiddenArrow)
//             Container(
//               padding: EdgeInsets.only(right: 10),
//               child: TriangleUpWidget(height: 10, width: 14),
//             ),
//           ClipRRect(
//               borderRadius: BorderRadius.circular(5),
//               child: Container(
//                 color: backgroundColor,
//                 width: width,
//                 height: height,
//                 child: widget,
//               ))
//         ],
//       ),
//     );
//   }
//   showDialog(
//       context: context,
//       barrierDismissible: false,
//       barrierColor: Colors.transparent,
//       builder: (context) {
//         return BasePopMenus(child: _buildMenusView());
//       });
// }

void showPopView({
  required BuildContext context,
  Widget? widget,
  double? top,
  double? left,
  double? right,
  double width = 100,
  double? height,
  Color backgroundColor = Colors.white,
  required Color barrierColor,
  bool hiddenArrow = false,
  bool barrierDismissible = true,
}) {
  RenderBox renderBox = context.findRenderObject() as RenderBox;
  Rect position = renderBox.localToGlobal(Offset.zero) & renderBox.size;
  if (top == null) {
    top = position.bottom;
  }
  if ((left == null && right == null)) {
    left = position.left - renderBox.size.width;
  }

  _buildMenusView() {
    return Material(
      type: MaterialType.transparency,
      child: Stack(
        children: [
          Positioned(
            top: top,
            left: left,
            right: right,
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.end,
              children: <Widget>[
                if (!hiddenArrow)
                  Container(
                    padding: EdgeInsets.only(right: 10),
                    child: TriangleUpWidget(height: 10, width: 14),
                  ),
                ClipRRect(
                    borderRadius: BorderRadius.circular(5),
                    child: Container(
                      color: backgroundColor,
                      width: width,
                      height: height,
                      child: widget,
                    ))
              ],
            ),
          ),
        ],
      ),
    );
  }

  int animalType = -1;

  showGeneralDialog(
    barrierDismissible: barrierDismissible,
    barrierLabel: 'menus',
    barrierColor: barrierColor,
    context: context,
    transitionDuration: Duration(milliseconds: 200),
    transitionBuilder: (BuildContext c, Animation<double> animation,
        Animation<double> secondaryAnimation, Widget child) {
      if (animalType == 0) {
        return FadeTransition(
          opacity: Tween(begin: 0.0, end: 1.0).animate(CurvedAnimation(
            parent: animation,
            curve: Curves.fastOutSlowIn,
          )),
          child: child,
        );
      } else if (animalType == 1) {
        return ScaleTransition(
          scale: Tween(
            begin: 0.0,
            end: 1.0,
          ).animate(
            CurvedAnimation(parent: animation, curve: Curves.fastOutSlowIn),
          ),
          child: child,
        );
      } else if (animalType == 2) {
        return RotationTransition(
          turns: Tween(
            begin: 0.0,
            end: 1.0,
          ).animate(
            CurvedAnimation(
              parent: animation,
              curve: Curves.fastOutSlowIn,
            ),
          ),
          child: child,
        );
      } else if (animalType == 3) {
        return SlideTransition(
          position: Tween<Offset>(
            begin: Offset(-1.0, 0.0),
            end: Offset(0.0, 0.0),
          ).animate(
            CurvedAnimation(
              parent: animation,
              curve: Curves.fastOutSlowIn,
            ),
          ),
          child: child,
        );
      }
      return SizeTransition(
        sizeFactor: Tween<double>(begin: 0.1, end: 1.0).animate(
          CurvedAnimation(
            parent: animation,
            curve: Curves.linear,
          ),
        ),
        child: child,
      );
    },
    pageBuilder: (BuildContext c, Animation<double> animation,
        Animation<double> secondaryAnimation) {
      return _buildMenusView();
    },
  );
}

class BasePopMenus extends Dialog {
  BasePopMenus({
    Key? key,
    required this.child,
  }) : super(key: key);

  final Widget child;

  @override
  Widget build(BuildContext context) {
    return Material(
      type: MaterialType.transparency,
      child: Stack(
        fit: StackFit.expand,
        children: <Widget>[
          GestureDetector(onTap: () => Navigator.pop(context)),
          child,
        ],
      ),
    );
  }
}

class TriangleUpPainter extends CustomPainter {
  Color? color; //填充颜色
  late Paint _paint; //画笔
  late Path _path; //绘制路径
  double? angle; //角度

  TriangleUpPainter(Color color) {
    _paint = Paint()
      ..strokeWidth = 1.0 //线宽
      ..color = color
      ..isAntiAlias = true;
    _path = Path();
  }

  @override
  void paint(Canvas canvas, Size size) {
    final baseX = size.width;
    final baseY = size.height;
    //起点
    _path.moveTo(baseX * 0.5, 0);
    _path.lineTo(baseX, baseY);
    _path.lineTo(0, baseY);
    canvas.drawPath(_path, _paint);
  }

  @override
  bool shouldRepaint(CustomPainter oldDelegate) {
    return false;
  }
}

class TriangleUpWidget extends StatefulWidget {
  double height;
  double width;
  Color color;

  TriangleUpWidget({
    Key? key,
    this.height = 14,
    this.width = 16,
    this.color = Colors.white,
  }) : super(key: key);

  @override
  CoreTriangleState createState() => CoreTriangleState();
}

class CoreTriangleState extends State<TriangleUpWidget> {
  @override
  Widget build(BuildContext context) {
    return Container(
        height: widget.height,
        width: widget.width,
        child: CustomPaint(
          painter: TriangleUpPainter(widget.color),
        ));
  }
}


lib/jd_drop_down_menu.dart
library jd_drop_down_menu;

import 'package:flutter/material.dart';

typedef PageBuilder = Widget Function(BuildContext context);

//用于导航下显示单个下拉菜单

void showDropdownMenu({
  required BuildContext context,
  required PageBuilder pageBuilder,
  required Rect position,
  double menuHeight = 100,
  bool barrierDismissible = true,
  Duration transitionDuration = const Duration(milliseconds: 300),
}) {
  assert(position != null);
  Navigator.push(
      context,
      _JDDropDownMenuRoute(
        pageBuilder: pageBuilder,
        position: position,
        menuHeight: menuHeight,
        barrierDismissible: barrierDismissible,
        transitionDuration: transitionDuration,
      ));
}

class _JDDropDownMenuRoute extends PopupRoute {
  final Rect? position;
  final double? menuHeight;

  _JDDropDownMenuRoute({
    required PageBuilder pageBuilder,
    this.position,
    this.menuHeight,
    Color barrierColor = const Color(0x80000000),
    bool barrierDismissible = true,
    Duration? transitionDuration,
  })  : _pageBuilder = pageBuilder,
        _barrierColor = barrierColor,
        _barrierDismissible = barrierDismissible,
        _transitionDuration = transitionDuration;

  final PageBuilder _pageBuilder;
  final Color _barrierColor;
  final bool _barrierDismissible;
  final Duration? _transitionDuration;
  @override
  Color? get barrierColor => null;

  @override
  bool get barrierDismissible => false;

  @override
  String? get barrierLabel => null;

  @override
  Duration get transitionDuration => _transitionDuration!;

  // @override
  // bool get opaque => true;

  @override
  Widget buildPage(BuildContext context, Animation<double> animation,
      Animation<double> secondaryAnimation) {
    Widget child = _pageBuilder(context);

    return CustomSingleChildLayout(
      delegate: _JDDropDownMenuRouteLayout(
          position: position, menuHeight: menuHeight),
      child: Container(
        color: _barrierColor,
        child: Column(
          children: [
            SizeTransition(
              sizeFactor:
                  Tween<double>(begin: 0.0, end: 1.0).animate(animation),
              child: Container(
                child: child,
                height: menuHeight,
              ),
            ),
            Expanded(
              child: GestureDetector(
                onTap: () {
                  if (_barrierDismissible) {
                    Navigator.pop(context);
                  }
                },
                child: Container(
                  color: Colors.transparent,
                ),
              ),
            )
          ],
        ),
      ),
    );

    // return Container(
    //   color: Colors.transparent,
    //   child: Column(
    //     children: [
    //       Container(
    //         margin: EdgeInsets.only(top: position.bottom),
    //         child: SizeTransition(
    //           sizeFactor:
    //               Tween<double>(begin: 0.0, end: 1.0).animate(animation),
    //           child: Container(
    //             child: child,
    //             height: menuHeight,
    //           ),
    //         ),
    //       ),
    //       Expanded(
    //         child: GestureDetector(
    //           onTap: () {
    //             if (_barrierDismissible) {
    //               Navigator.pop(context);
    //             }
    //           },
    //           child: Container(
    //             color: _barrierColor,
    //           ),
    //         ),
    //       )
    //     ],
    //   ),
    // );
  }
}

class _JDDropDownMenuRouteLayout extends SingleChildLayoutDelegate {
  final Rect? position;
  final double? menuHeight;
  _JDDropDownMenuRouteLayout({this.position, this.menuHeight});

  @override
  BoxConstraints getConstraintsForChild(BoxConstraints constraints) {
    return BoxConstraints.loose(
        Size(constraints.maxWidth, constraints.maxHeight));
  }

  @override
  Offset getPositionForChild(Size size, Size childSize) {
    return Offset(0, position!.bottom);
  }

  @override
  bool shouldRelayout(SingleChildLayoutDelegate oldDelegate) {
    return true;
  }
}


lib/jd_dropdowm_menu_widget.dart
import 'package:flutter/material.dart';
import 'package:flutter/widgets.dart';

/// @author jd

typedef JDDropdownMenuClick = void Function(int index);
typedef JDDropdownMenuDismiss = void Function();

//用于多级下拉菜单

enum JDDropdownMenuWidgetStyle {
  style1, //自适应
  style2, //等分
}

class JDDropdownMenuItem {
  final Widget title; //菜单点击区域
  final Widget menu; //菜单下拉内容区域
  final double maxHeight; //内容区域高度

  JDDropdownMenuItem({
    required this.title,
    required this.menu,
    this.maxHeight = 100,
  });
}

//Controller
class JDDropdownMenuController extends ChangeNotifier {
  bool isShow = false;
  int menuIndex = 0;

  void show(int index) {
    isShow = true;
    menuIndex = index;
    notifyListeners();
  }

  void hide() {
    isShow = false;
    notifyListeners();
  }
}

class JDDropdownMenuWidget extends StatefulWidget {
  final List<JDDropdownMenuItem> items;
  final JDDropdownMenuClick? click;
  final JDDropdownMenuDismiss? dismiss;
  final JDDropdownMenuWidgetStyle? itemStyle;
  final JDDropdownMenuController? controller;
  final double height;
  JDDropdownMenuWidget({
    required this.items,
    this.height = 40,
    this.click,
    this.dismiss,
    this.itemStyle = JDDropdownMenuWidgetStyle.style1,
    this.controller,
  });

  @override
  _JDDropdownMenuWidgetState createState() => _JDDropdownMenuWidgetState();
}

class _JDDropdownMenuWidgetState extends State<JDDropdownMenuWidget>
    with SingleTickerProviderStateMixin {
  OverlayEntry? _entry;
  late Animation _animation;
  late AnimationController _animationController;

  @override
  void initState() {
    _animationController = AnimationController(
        duration: Duration(
          milliseconds: 200,
        ),
        vsync: this);
    _animation = Tween(begin: 0.0, end: 1.0).animate(_animationController);
    _animationController.addListener(() {
      if (_animationController.status == AnimationStatus.dismissed) {
        _entry!.remove();
        _entry = null;
        if (widget.dismiss != null) {
          widget.dismiss!();
        }
      }
    });
    widget.controller?.addListener(() {
      if (widget.controller!.isShow) {
        JDDropdownMenuItem item = widget.items[widget.controller!.menuIndex];
        showMenu(context, item);
      } else {
        dismiss();
      }
    });
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return SizedBox(
      height: widget.height,
      child: Row(
        children: widget.items
            .map(
              (e) => menuItemWidget(e),
            )
            .toList(),
      ),
    );
  }

  Widget menuItemWidget(JDDropdownMenuItem item) {
    Widget menu = Builder(
      builder: (BuildContext c) => GestureDetector(
        onTap: () {
          showMenu(c, item);
        },
        child: item.title,
      ),
    );
    if (widget.itemStyle == JDDropdownMenuWidgetStyle.style2) {
      menu = Expanded(
        child: menu,
      );
    }
    return menu;
  }

  void showMenu(BuildContext context, JDDropdownMenuItem item) {
    widget.controller?.isShow = true;
    if (widget.click != null) {
      widget.click!(widget.items.indexOf(item));
    }
    RenderBox renderBox = context.findRenderObject() as RenderBox;
    Rect position = renderBox.localToGlobal(Offset.zero) & renderBox.size;
    if (_entry != null) {
      _entry!.remove();
      _entry = null;
    }

    _entry = OverlayEntry(
      builder: (BuildContext context) => Positioned.fill(
        top: position.bottom,
        child: Material(
          color: Colors.transparent,
          child: Column(
            children: [
              SizeTransition(
                sizeFactor: _animation as Animation<double>,
                child: Container(
                  child: item.menu,
                  height: item.maxHeight,
                ),
              ),
              Expanded(
                child: GestureDetector(
                  onTap: () {
                    dismiss();
                  },
                  child: FadeTransition(
                    opacity: _animationController,
                    child: Container(
                      color: const Color(0x80000000),
                    ),
                  ),
                ),
              )
            ],
          ),
        ),
      ),
    );

    Overlay.of(context)!.insert(_entry!);
    _animationController.forward();
  }

  void dismiss() {
    widget.controller?.isShow = false;
    if (_entry != null) {
      _animationController.reverse();
    }
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:jd_dropdowm_menu_widget/jd_drop_down_menu.dart';
import 'package:jd_dropdowm_menu_widget/jd_dropdowm_menu_widget.dart';
import 'package:jd_dropdowm_menu_widget/jd_pop_widget.dart';

import 'demo/city_widget.dart';
import 'demo/menu_list_widget.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        // This is the theme of your application.
        //
        // Try running your application with "flutter run". You'll see the
        // application has a blue toolbar. Then, without quitting the app, try
        // changing the primarySwatch below to Colors.green and then invoke
        // "hot reload" (press "r" in the console where you ran "flutter run",
        // or simply save your changes to "hot reload" in a Flutter IDE).
        // Notice that the counter didn't reset back to zero; the application
        // is not restarted.
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  // This widget is the home page of your application. It is stateful, meaning
  // that it has a State object (defined below) that contains fields that affect
  // how it looks.

  // This class is the configuration for the state. It holds the values (in this
  // case the title) provided by the parent (in this case the App widget) and
  // used by the build method of the State. Fields in a Widget subclass are
  // always marked "final".

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  String title1 = '城市';
  String title2 = '品牌';
  GlobalKey _key1 = GlobalKey();

  JDDropdownMenuController controller = JDDropdownMenuController();
  @override
  Widget build(BuildContext context) {
    // This method is rerun every time setState is called, for instance as done
    // by the _incrementCounter method above.
    //
    // The Flutter framework has been optimized to make rerunning build methods
    // fast, so that you can just rebuild anything that needs updating rather
    // than having to individually change instances of widgets.
    return Scaffold(
      appBar: AppBar(
        // Here we take the value from the MyHomePage object that was created by
        // the App.build method, and use it to set our appbar title.
        title: Text(widget.title),
        actions: [
          TextButton(
              key: _key1,
              onPressed: () {
                RenderBox renderBox = _key1.currentContext.findRenderObject();
                Rect position =
                    renderBox.localToGlobal(Offset.zero) & renderBox.size;
                showDropdownMenu(
                  position: position,
                  context: context,
                  pageBuilder: (c1) => Container(
                    color: Colors.orange,
                  ),
                );
              },
              child: Text(
                '下拉菜单',
                style: TextStyle(
                  color: Colors.white,
                ),
              ))
        ],
      ),
      body: Column(
        children: [
          Container(
            margin: EdgeInsets.only(
              top: 10,
            ),
            child: JDDropdownMenuWidget(
              controller: controller,
              itemStyle: JDDropdownMenuWidgetStyle.style2,
              click: (int index) {
                print('$index');
              },
              items: [
                JDDropdownMenuItem(
                  maxHeight: 300,
                  title: Container(
                    alignment: Alignment.topCenter,
                    margin: EdgeInsets.only(
                      left: 10,
                      right: 10,
                    ),
                    child: Text(title1),
                  ),
                  menu: CityWidget(
                    click: (String selectedValue) {
                      setState(() {
                        title1 = selectedValue;
                        controller.hide();
                      });
                      print('选择的城市为:$selectedValue');
                    },
                  ),
                ),
                JDDropdownMenuItem(
                  maxHeight: 300,
                  title: Container(
                    alignment: Alignment.topCenter,
                    margin: EdgeInsets.only(
                      left: 10,
                      right: 10,
                    ),
                    child: Text(title2),
                  ),
                  menu: MenuListWidget(
                    selectedValue: title2,
                    click: (String value) {
                      setState(() {
                        title2 = value;
                        controller.hide();
                      });
                    },
                  ),
                ),
                JDDropdownMenuItem(
                  title: Container(
                    alignment: Alignment.topCenter,
                    margin: EdgeInsets.only(
                      left: 10,
                      right: 10,
                    ),
                    child: Text('菜单3'),
                  ),
                  menu: Container(
                    color: Colors.blue,
                  ),
                ),
              ],
            ),
          ),
          Container(
            child: Builder(
              builder: (c) {
                return TextButton(
                    onPressed: () {
                      RenderBox renderBox = c.findRenderObject();
                      Rect position =
                          renderBox.localToGlobal(Offset.zero) & renderBox.size;
                      showDropdownMenu(
                        position: position,
                        context: c,
                        pageBuilder: (c1) => Container(
                          color: Color(0xffff0000),
                        ),
                      );
                    },
                    child: Text('Menu1'));
              },
            ),
          ),
          Container(
            child: Builder(
              builder: (c) {
                return TextButton(
                    onPressed: () {
                      showPop(
                        backgroundColor: Colors.red,
                        barrierColor: const Color(0x00000000),
                        context: c,
                        items: [
                          Text(
                            '个人信息',
                            style: TextStyle(color: Colors.white, fontSize: 14),
                          ),
                          Text(
                            '退出',
                            style: TextStyle(color: Colors.white, fontSize: 14),
                          ),
                        ],
                      );
                    },
                    child: Text('Menu2'));
              },
            ),
          ),
        ],
      ), // This trailing comma makes auto-formatting nicer for build methods.
    );
  }
}
